src/Compilers/CSharp/Portable/Binder/WithUsingNamespacesAndTypesBinder.cs

```diff
@@ -127,6 +127,33 @@ internal override bool SupportsExtensionMethods
             }
         }
 
+        internal override bool SupportsExtensionTypes => true;
+
+        internal override void GetImplicitExtensionTypes(ArrayBuilder<TypeSymbol> extensions, Binder originalBinder)
+        {
+            Debug.Assert(extensions.Count == 0);
+
+            // PROTOTYPE test this flag (see TestUnusedExtensionMarksImportsAsUsed)
+            bool callerIsSemanticModel = originalBinder.IsSemanticModelBinder;
+
+            foreach (var nsOrType in this.GetUsings(basesBeingResolved: null))
+            {
+                if (nsOrType.NamespaceOrType is NamespaceSymbol ns)
+                {
+                    var count = extensions.Count;
+                    ns.GetImplicitExtensionTypes(extensions);
+                    // If we found any extension types, then consider this using as used.
+                    // PROTOTYPE consider refining this logic
+                    if (extensions.Count != count)
+                    {
+                        MarkImportDirective(nsOrType.UsingDirectiveReference, callerIsSemanticModel);
+                    }
+                }
```

Comment:
>} [](http://example.com/codeflow?start=16&length=1)

I think for ```using static SomeType```, we should look for extension types declared within ```SomeType```. This is going to be consistent with how we look for extension methods and how we look for extension types in an enclosing type declaration. Types are never instance members, therefore, it is appropriate to look for them in ```using static```. In fact we do look for types in a scenario like that:
```
using static C1;

C2 x = null;

class C1
{
    public class C2{}
}
```

Note, this is different than the open design question: "What to do with ```Extension``` type itself when we run into a ```using static Extension```."

Reply:
That makes sense. I captured some notes in the next spec update ([https://github.com/dotnet/csharplang/pull/7179](https://github.com/dotnet/csharplang/pull/7179)).